boto/s3/bucket.py

```diff
@@ -879,15 +884,33 @@ def delete_keys2(headers):
             data += u"</Delete>"
             if count <= 0:
                 return False  # no more
-            data = data.encode('utf-8')
-            fp = BytesIO(data)
-            md5 = boto.utils.compute_md5(fp)
+            md5 = boto.utils.compute_md5(StringIO(data))
             headers['Content-MD5'] = md5[1]
             headers['Content-Type'] = 'text/xml'
+            if not isinstance(data, bytes):
+                data = data.encode('utf-8')
```

Comment:
When is the data not a utf-8 string at this point? Ie, I am wondering what I am missing as to why you are checking if it is already bytes when you are just converting it at this point? The older code also used to just directly encode it to bytes.

Reply:
Though have not checked all configuration APIs callers(from toms3/models3 and others) which support checksum feature, the passing data could be 'string' or 'bytes' depending on the caller. I guess most of data will be 'string' but just want to have data as 'bytes' here before calling cal_checksum method as its data is required as 'bytes'. So this may not be mostly necessary but just want to check it before calculating the checksum for each configuration API just in case...

Reply:
The `data` variable in this routine is always a Unicode string at this point as it is being composed using `u"</Delete>"` etc which you can see just above. That is why you don't need to check that it is not bytes before converting to bytes.

Reply:
Ok. Reviewed again and made some changes to convert without checking if data is obviously a string type. https://github.com/cloudian/boto/pull/42/commits/01266d181044b7a037ac4b10fc4fa750ffc23ce2